tests/test_branch.rs

```diff
@@ -0,0 +1,415 @@
+extern crate narm;
+mod common;
+
+use common::*;
+use narm::narmvm::*;
+
+/*
+
+Unit test for Branch operators
+
+Included varieties:
+
+B <label> T2            - Branch unconditionally by label, jump size restricted to signed 11 bits
+B<C> <label> T1         - Branch conditionally by label, jump size restricted to signed 8 bits
+BX <Rm> T1              - Branch by register
+BL <label> T1           - Branch by label, set link register, jump size restricted to even signed 25 bits (?????)
+BLX <Rm> T1             - Branch by register, set link register
+
+General test cases:
+
+- Branch forward
+- Branch backward
+- Branch far forward (Causing memory error)
+- Branch far backward (Causing memory error)
+
+Special test case for BLX <Rm> T1:
+
+- Branch and then branch back using address saved in link register
+
+Special test case for B<C>:
+
+- Test all different conditions
+
+The reference for these tests is currently official documentations and a QEMU-based VM
+TODO: Test against a hardware Cortex-M0 to make sure it's actually up to spec?
+
+*/
+
+// String representation of ops for use in debug output
+const OPCODES: &'static [&'static str] = &[
+    "B <label> T2",
+    "B<C> <label> T1",
+    "BX <Rm> T1",
+    "BL <label> T1 32bit",
+    "BLX <Rm> T1",
+];
+
+// Simple constant for number of opcodes tested in this file
+const NUM_OPCODES: &'static usize = &5;
+
+// Branch forward
+#[test]
+pub fn test_branch_forward() {
+    println!("\n>>> Branch ops test case: Branch forward \n");
+
+    // Arrays holding instances of VMs and matching state structs
+    let mut vms: [NarmVM; *NUM_OPCODES] = Default::default();
+    let mut vm_states: [VMState; *NUM_OPCODES] = Default::default();
+
+    // Tell macros which op varieties are tested in this function
+    let applicable_op_ids = vec![0, 1, 2, 3, 4];
+
+    // Common pre-execution state
+    common_state!(
+        applicable_op_ids,
+        vm_states.r[1] = Some(ASM_ENTRY + 2 * OP_SIZE + THUMBS_MODE)
```

Comment:
This could likely be made less noisy by computing this address with a function like `test_code_offset(2 * OP_SIZE)` where code_offset is something like
```
fn test_code_offset(offset: u32) -> u32{
    ASM_ENTRY + THUMBS_MODE + offset
}
```

A simple change, but even with constants it's quite noisey about why THUMBS_MODE is needed and what ASM_ENTRY is

Reply:
Good point, I'll do that! Probably as a macro though, I've walked too far down that path to turn back now.

Reply:
Just be careful not to overuse macros where they aren't needed

Reply:
Yeah, I'm working on when to use what.